src/System.Net.Security/src/System/Net/Security/SslApplicationProtocol.cs

```diff
@@ -2,139 +2,111 @@
+        public static readonly SslApplicationProtocol Http11 = new SslApplicationProtocol(s_http11Utf8, copy: false);
+
+        private readonly byte[] _readOnlyProtocol;
 
         internal SslApplicationProtocol(byte[] protocol, bool copy)
         {
-            if (protocol == null)
-            {
-                throw new ArgumentNullException(nameof(protocol));
-            }
+            Debug.Assert(protocol != null);
 
             // RFC 7301 states protocol size <= 255 bytes.
             if (protocol.Length == 0 || protocol.Length > 255)
             {
                 throw new ArgumentException(SR.net_ssl_app_protocol_invalid, nameof(protocol));
             }
 
-            if (copy)
-            {
-                byte[] temp = new byte[protocol.Length];
-                Array.Copy(protocol, 0, temp, 0, protocol.Length);
-                _readOnlyProtocol = new ReadOnlyMemory<byte>(temp);
-            }
-            else
-            {
-                _readOnlyProtocol = new ReadOnlyMemory<byte>(protocol);
-            }
+            _readOnlyProtocol = copy ?
+                protocol.AsSpan().ToArray() :
+                protocol;
         }
 
-        public SslApplicationProtocol(byte[] protocol) : this(protocol, true) { }
-
-        public SslApplicationProtocol(string protocol) : this(s_utf8.GetBytes(protocol), copy: false) { }
-
-        public ReadOnlyMemory<byte> Protocol
+        public SslApplicationProtocol(byte[] protocol) :
+            this(protocol ?? throw new ArgumentNullException(nameof(protocol)), copy: true)
         {
-            get => _readOnlyProtocol;
         }
 
-        public bool Equals(SslApplicationProtocol other)
+        public SslApplicationProtocol(string protocol) :
+            this(s_utf8.GetBytes(protocol ?? throw new ArgumentNullException(nameof(protocol))), copy: false)
         {
-            if (_readOnlyProtocol.Length != other._readOnlyProtocol.Length)
-                return false;
-
-            return (_readOnlyProtocol.IsEmpty && other._readOnlyProtocol.IsEmpty) ||
-                _readOnlyProtocol.Span.SequenceEqual(other._readOnlyProtocol.Span);
         }
 
-        public override bool Equals(object obj)
-        {
-            if (obj is SslApplicationProtocol protocol)
-            {
-                return Equals(protocol);
-            }
+        public ReadOnlyMemory<byte> Protocol => _readOnlyProtocol;
 
-            return false;
-        }
+        public bool Equals(SslApplicationProtocol other) =>
+            ((ReadOnlySpan<byte>)_readOnlyProtocol).SequenceEqual(other._readOnlyProtocol);
+
+        public override bool Equals(object obj) => obj is SslApplicationProtocol protocol && Equals(protocol);
 
         public override int GetHashCode()
         {
-            if (_readOnlyProtocol.Length == 0)
+            byte[] arr = _readOnlyProtocol;
```

Comment:
since we are modifying it - should we convert this to use Marvin hashing? If this is perf critical and we only have two values perhaps we should just return `arr[1]` since they differ only on the second character

Reply:
> we only have two values

We don't actually; the ctors are public and support arbitrary byte arrays / strings being passed in.

Without evidence about needing a better hash code here or perf being a factor, I'm going to leave the algorithm as is.